test/gateway/eip20_cogateway/revert_redeem.js

```diff
@@ -0,0 +1,363 @@
+      nonce: new BN(1),
+      hashLock: hashLockObj.l,
+      unlockSecret: hashLockObj,
+      redeemer: accounts[2],
+      intentHash: web3.utils.sha3("dummy"),
+      penalty: constructorParams.bounty.muln(1.5),
+    };
+
+    redeemParams.messageHash = await eip20CoGateway.setMessage.call(
+      redeemParams.intentHash,
+      redeemParams.nonce,
+      redeemParams.gasPrice,
+      redeemParams.gasLimit,
+      redeemParams.redeemer,
+      redeemParams.hashLock,
+    );
+
+    await eip20CoGateway.setMessage(
+      redeemParams.intentHash,
+      redeemParams.nonce,
+      redeemParams.gasPrice,
+      redeemParams.gasLimit,
+      redeemParams.redeemer,
+      redeemParams.hashLock,
+    );
+
+    await eip20CoGateway.setRedeem(
+      redeemParams.messageHash,
+      redeemParams.beneficiary,
+      redeemParams.amount,
+    );
+
+    await eip20CoGateway.setOutboxStatus(
+      redeemParams.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    // Set co-gateway to owner so that increase supply can be called.
+    await utilityToken.setCoGatewayAddress(constructorParams.owner);
+
+    // Send redeem amount to co-gateway.
+    await  utilityToken.increaseSupply(
+      eip20CoGateway.address,
+      redeemParams.amount,
+      { from: constructorParams.owner },
+    );
+
+    // Send bounty to co-gateway.
+    await web3.eth.sendTransaction({
+      to: eip20CoGateway.address,
+      from: constructorParams.owner,
+      value: constructorParams.bounty,
+    });
+
+    await utilityToken.setCoGatewayAddress(eip20CoGateway.address);
+
+  });
+
+  it('should fail when message hash is zero', async function () {
+
+    await Utils.expectRevert(
+      eip20CoGateway.revertRedeem(
+        zeroBytes,
+        { from: redeemParams.redeemer, value: redeemParams.penalty },
+      ),
+      'Message hash must not be zero.',
+    );
+
+  });
+
+  it('should fail when message is not declared', async function () {
+
+    await Utils.expectRevert(
+      eip20CoGateway.revertRedeem(
+        web3.utils.sha3("message_hash"),
+        { from: redeemParams.redeemer, value: redeemParams.penalty },
+      ),
+      'RedeemIntentHash must not be zero.',
+    );
+
+  });
+
+  it('should fail when msg.sender is not redeemer address', async function () {
+
+    await Utils.expectRevert(
+      eip20CoGateway.revertRedeem(
+        redeemParams.messageHash,
+        { from: constructorParams.owner, value: redeemParams.penalty },
+      ),
+      'Only redeemer can revert redeem.',
+    );
+
+  });
+
+  it('should fail when msg.value is less than penalty amount', async function () {
+
+    await Utils.expectRevert(
+      eip20CoGateway.revertRedeem(
+        redeemParams.messageHash,
+        { from: redeemParams.redeemer, value: constructorParams.bounty.muln(1.25) },
+      ),
+      'msg.value must match the penalty amount.',
+    );
+
+  });
+
+  it('should fail when msg.value is greater than penalty amount ', async function () {
```

Comment:
Not sure, if we should allow for a greater penalty amount. 

@schemar  what do you think?

Reply:
What would happen to the tokens? They would be stuck in the gateway, right? Or do you suggest that I can send a greater value and then all of that would be the penalty?
It seems that the existing check is nice to prevent user errors. What if I accidentally send 10 times the amount I intended to send?
💰💥